kube/configmap.go

```diff
@@ -1,135 +1,125 @@
 package kube
 
 import (
-	"errors"
-	"fmt"
-	"time"
+	"context"
+	"net"
 
 	log "github.com/sirupsen/logrus"
-	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/szuecs/kube-static-egress-controller/provider"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/kubernetes"
-	v1 "k8s.io/client-go/pkg/api/v1"
-)
-
-var (
-	ErrClosedChannel = errors.New("closed watch channel")
+	"k8s.io/client-go/tools/cache"
 )
 
 type ConfigMapWatcher struct {
-	client   kubernetes.Interface
-	ns       string
-	selector string
-	opts     meta_v1.ListOptions
-	quitCH   <-chan struct{}
+	client    kubernetes.Interface
+	namespace string
+	selector  fields.Selector
+	configs   chan<- provider.EgressConfig
 }
 
-func NewConfigMapWatcher(client kubernetes.Interface, ns, selector string, quitCH <-chan struct{}) (*ConfigMapWatcher, error) {
+func NewConfigMapWatcher(client kubernetes.Interface, namespace, selectorStr string, configs chan<- provider.EgressConfig) (*ConfigMapWatcher, error) {
+	selector, err := fields.ParseSelector(selectorStr)
+	if err != nil {
+		return nil, err
+	}
+
 	return &ConfigMapWatcher{
-		client:   client,
-		ns:       ns,
-		selector: selector,
-		opts:     meta_v1.ListOptions{LabelSelector: selector},
-		quitCH:   quitCH,
+		client:    client,
+		namespace: namespace,
+		selector:  selector,
+		configs:   configs,
 	}, nil
 }
 
-type Tuple struct {
-	Key   string
-	Value string
-}
+func (c *ConfigMapWatcher) Run(ctx context.Context) {
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = c.selector.String()
+				return c.client.CoreV1().ConfigMaps(c.namespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = c.selector.String()
+				return c.client.CoreV1().ConfigMaps(c.namespace).Watch(options)
+			},
+		},
+		&v1.ConfigMap{},
+		0, // skip resync
```

Comment:
Does the resync not provide the tool, that we now manually have added with the time.After() ?

Reply:
This would cause all configmaps to be resynced, resulting in `n` events. To simplify it I use the `time.After` to only have `1` event on every resync. This issue is still there when starting the controller. But I think it's ok for now.

Reply:
Sounds good